### PR TITLE
Lock the rows when taking the advisory lock with ActiveRecordWithLock adapter

### PR DESCRIPTION
**Pull request by ankithads**
For ActiveRecordWithLock adapter, we want to lock the rows when acquiring the advisory locks on the table to avoid all the workers trying to lock the same job.

Added metrics to observe the latency to find the job and job hit rate.

### CODE REVIEW

File: lib/que/sql.rb [hunk 1]
```diff
@@ -184,7 +184,7 @@ module Que
             AND retryable = true
             AND job_id >= $2
             ORDER BY priority, run_at, job_id
-            LIMIT 1
+            for update skip locked LIMIT 1
```
Comment (ameykusurkar) [inline lib/que/sql.rb:187]:
```suggestion
            FOR UPDATE SKIP LOCKED LIMIT 1
```

File: lib/que/adapters/active_record_with_lock.rb [hunk 2]
```diff
@@ -35,13 +47,22 @@ def execute(command, params = [])
       def lock_job_with_lock_database(queue, cursor)
         result = []
         loop do
-          result = Que.execute(:find_job_to_lock, [queue, cursor])
+          result = Que.transaction do
+            observe(nil, FindJobSecondsTotal, queue: queue) do
+              result = Que.execute(:find_job_to_lock, [queue, cursor])
```
Comment (ameykusurkar) [inline lib/que/adapters/active_record_with_lock.rb:52]: I don't think it's obvious why we do this, given that there is already an advisory lock. Let's explain in a comment why this improves performance when there are a large number of workers contending for work, and why this doesn't ensure race safety (hence we still need the advisory lock).

File: lib/que/adapters/active_record_with_lock.rb [hunk 3]
```diff
@@ -35,13 +47,22 @@ def execute(command, params = [])
       def lock_job_with_lock_database(queue, cursor)
         result = []
         loop do
-          result = Que.execute(:find_job_to_lock, [queue, cursor])
+          result = Que.transaction do
+            observe(nil, FindJobSecondsTotal, queue: queue) do
```
Comment (ameykusurkar) [inline lib/que/adapters/active_record_with_lock.rb:51]: Nit: let's put this outside the transaction?

File: spec/active_record_with_lock_spec_helper.rb [hunk 4]
```diff
@@ -4,7 +4,7 @@ class LockDatabaseRecord < ActiveRecord::Base
   establish_connection(
     adapter: "postgresql",
     host: ENV.fetch("LOCK_PGHOST", "localhost"),
-    user: ENV.fetch("LOCK_PGUSER", "postgres"),
+    user: ENV.fetch("LOCK_PGUSER", "ubuntu"),
```
Comment (mfmmq) [inline spec/active_record_with_lock_spec_helper.rb:7]: Why is this ubuntu ?  Is this something specific to CI specs?

File: lib/que/adapters/active_record_with_lock.rb [hunk 5]
```diff
@@ -63,6 +84,21 @@ def unlock_job(job_id)
           conn.execute("SELECT pg_advisory_unlock(#{job_id})")
         end
       end
+
+      def observe(metric, metric_duration, labels = {})
```
Comment (ameykusurkar) [inline lib/que/adapters/active_record_with_lock.rb:88]:
```suggestion
      def observe(count_metric: nil, duration_metric: nil, labels: {})
```
use named args with defaults here, as it's not always clear what the order of arguments mean, and why we're passing in nil

File: spec/active_record_with_lock_spec_helper.rb (hunk 4 above)
Comment (ameykusurkar) [inline spec/active_record_with_lock_spec_helper.rb:7]: Curious, what's the reason for this change?

File: spec/active_record_with_lock_spec_helper.rb [hunk 6]
```diff
@@ -28,3 +28,11 @@ def active_record_with_lock_adapter_connection
...
   )
 end
+
+RSpec.configure do |config|
+  if ENV["ADAPTER"] == "ActiveRecordWithLock"
+    config.filter_run_including :active_record_with_lock
+  else
+    config.filter_run_excluding :active_record_with_lock
+  end
+end
```
Comment (ameykusurkar) [inline spec/active_record_with_lock_spec_helper.rb:38]: I think it can be unintuitive that only some specs run. Given that this is to test adapter spec functionality, could we just add an RSpec for that class?

File: lib/que/adapters/active_record_with_lock.rb [hunk 7]
```diff
@@ -35,13 +47,22 @@ def execute(command, params = [])
...
+              result = Que.execute(:find_job_to_lock, [queue, cursor])
+            end
 
-          break if result.empty?
+            return result if result.empty?
 
-          cursor = result.first["job_id"]
-          break if pg_try_advisory_lock?(cursor)
+            cursor = result.first["job_id"]
+            job_locked = pg_try_advisory_lock?(cursor)
```
Comment (mfmmq) [inline lib/que/adapters/active_record_with_lock.rb:58]: Do we want to include the SQL query here in the metric for the time it takes to find a job ?

Comment (mfmmq) [reply lib/que/adapters/active_record_with_lock.rb:58]: Nit + non blocking for this PR: also, if you unify the metric labels you might be able to use `Hesiod.register_duration_counters` and `Hesiod.observe` as convenience wrappers

File: lib/que/adapters/active_record_with_lock.rb [hunk 8]
```diff
@@ -35,13 +47,22 @@ def execute(command, params = [])
...
+
+            observe(FindJobHitTotal, nil, { queue: queue, job_hit: job_locked })
+            return result if job_locked
+          end
+          break if result
```
Comment (ameykusurkar) [inline lib/que/adapters/active_record_with_lock.rb:63]: Nit: a combination of returns and breaks can be a bit hard to follow. Is it possible to avoid this break such that all result cases are handled by early returns?

File: spec/active_record_with_lock_spec_helper.rb [hunk 9]
```diff
@@ -28,3 +28,11 @@ def active_record_with_lock_adapter_connection
     lock_connection_pool: LockDatabaseRecord.connection_pool,
   )
 end
+
+RSpec.configure do |config|
+  if ENV["ADAPTER"] == "ActiveRecordWithLock"
```
Comment (mfmmq) [inline spec/active_record_with_lock_spec_helper.rb:33]: What's the reason for gating this test? Also, if we're going to gate it do we need to document it/specify it in CI ?

**Review by benk-gc — COMMENTED** [review]
A few comments on the control flow. I have a meta-point on the overall Que philosophy, since we're now breaking one of the assertions in the README:

> Workers don't block each other when trying to lock jobs, as often occurs with "SELECT FOR UPDATE"-style locking.

If we're now breaking this assertion for these jobs we should make this clear and justify why we're trading this off for the performance hit.

File: lib/que/adapters/active_record_with_lock.rb (hunk 7 above)
Comment (ankithads) [reply lib/que/adapters/active_record_with_lock.rb:58]: time to find a job is in line 51. Locking metric shouldn't take much time. Not required to track as of now. 
There is already a metric within Locker to track this

File: spec/integration/integration_spec.rb [hunk 10]
```diff
@@ -119,6 +119,20 @@ def wait_for_jobs_to_be_worked(timeout: 10)
       expect(User.count).to eq(3)
       expect(User.all.map(&:name).sort).to eq(%w[alice bob charlie])
     end
+
+    it "increments the metrics", :active_record_with_lock do
```
[the remark this answers is not in the document]
Comment (ankithads) [reply spec/integration/integration_spec.rb]: updated

File: lib/que/adapters/active_record_with_lock.rb [hunk 11]
```diff
@@ -3,6 +3,19 @@
 module Que
   module Adapters
     class ActiveRecordWithLock < Que::Adapters::ActiveRecord
+      class NoLockableJobs < StandardError; end
```
Comment (ameykusurkar) [inline lib/que/adapters/active_record_with_lock.rb:6]: Is this leftover code? I don't see it being used in this class

File: lib/que/adapters/active_record_with_lock.rb [hunk 12]
```diff
@@ -32,17 +45,25 @@ def execute(command, params = [])
...
 
-          cursor = result.first["job_id"]
-          break if pg_try_advisory_lock?(cursor)
+              cursor = job_to_lock.first["job_id"]
+              job_locked = pg_try_advisory_lock?(cursor)
+
+              observe(count_metric: FindJobHitTotal, labels: { queue: queue, job_hit: job_locked })
+              return job_to_lock if job_locked
+            end
+            return locked_job if locked_job
```
Comment (ameykusurkar) [inline lib/que/adapters/active_record_with_lock.rb:64]: When is this case hit? Looks like there are only two cases:
> locks a job successfully

on line 64

> or determines that there are no jobs to process.

line 56

Comment (ankithads) [reply lib/que/adapters/active_record_with_lock.rb:64]: yeah. This code is not reachable. Will remove this

File: spec/lib/que/adapters/active_record_with_lock_spec.rb [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+        example.run
+        Que.adapter = old_adapter
+      else
+        example.run
+      end
+    end
+  end
+
+  let(:lock_connection_pool) do
+    return LockDatabaseRecord.connection_pool if ENV["ADAPTER"] == "ActiveRecordWithLock"
```
Comment (ameykusurkar) [inline spec/lib/que/adapters/active_record_with_lock_spec.rb:25]: Now have separate tests for the two adapters, I think we should use shared specs to define "Que behaviour" for all adapters, so that we only run those per adapter rather than all the specs - this feels a bit clearer than controlling this through env vars.

Non-blocking, can be a clean-up later

**Review by ameykusurkar — APPROVED** [review]

File: spec/lib/que/adapters/active_record_with_lock_spec.rb (hunk 13 above)
Comment (ankithads) [reply spec/lib/que/adapters/active_record_with_lock_spec.rb:25]: Since this is required only for this Adapter I think adding a filter would be better. It gets bit tricky because of the 2 database here.
Rest all the specs are run against specific adapter.

Comment (ankithads) [reply spec/lib/que/adapters/active_record_with_lock_spec.rb]: I have updated this to exclude all the files in adapter spec directory except for the adapter spec
